boards/beagle/beagleconnect_freedom/board_antenna.c

```diff
@@ -1,116 +1,119 @@
-/* SPDX-License-Identifier: Apache-2.0
- *
+/*
  * Copyright (c) 2021 Florin Stancu
- * Copyright (c) 2021 Jason Kridner, BeagleBoard.org Foundation
+ * Copyright (c) 2024 Ayush Singh <ayush@beagleboard.org>
  *
+ * SPDX-License-Identifier: Apache-2.0
  */
 
 /*
  * Implements the RF driver callback to configure the on-board antenna
  * switch.
  */
 
-#include <zephyr/kernel.h>
-#include <zephyr/device.h>
+#define DT_DRV_COMPAT skyworks_sky13317
+
 #include <zephyr/init.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/pinctrl.h>
 
 #include <ti/drivers/rf/RF.h>
-#include <driverlib/gpio.h>
-#include <driverlib/ioc.h>
 #include <driverlib/rom.h>
+#include <driverlib/interrupt.h>
 
-/* DIOs for RF antenna paths */
-#define BOARD_RF_HIGH_PA   29		/* TODO: pull from DT */
-#define BOARD_RF_SUB1GHZ   30		/* TODO: pull from DT */
+/* custom pinctrl states for the antenna mux */
+#define PINCTRL_STATE_ANT_SUBG    1
+#define PINCTRL_STATE_ANT_SUBG_PA 2
 
-static void board_cc13xx_rf_callback(RF_Handle client, RF_GlobalEvent events,
-		void *arg);
+#define BOARD_ANT_GPIO_PA   0
+#define BOARD_ANT_GPIO_SUBG 1
 
+static int board_antenna_init(const struct device *dev);
+static void board_cc13xx_rf_callback(RF_Handle client, RF_GlobalEvent events, void *arg);
 
 const RFCC26XX_HWAttrsV2 RFCC26XX_hwAttrs = {
-	.hwiPriority        = INT_PRI_LEVEL7,
-	.swiPriority        = 0,
+	.hwiPriority = INT_PRI_LEVEL7,
+	.swiPriority = 0,
 	.xoscHfAlwaysNeeded = true,
 	/* RF driver callback for custom antenna switching */
 	.globalCallback = board_cc13xx_rf_callback,
 	/* Subscribe to events */
-	.globalEventMask = (RF_GlobalEventRadioSetup |
-			RF_GlobalEventRadioPowerDown),
+	.globalEventMask = (RF_GlobalEventRadioSetup | RF_GlobalEventRadioPowerDown),
 };
 
+PINCTRL_DT_INST_DEFINE(0);
+DEVICE_DT_INST_DEFINE(0, board_antenna_init, NULL, NULL, NULL, POST_KERNEL,
+		      CONFIG_BOARD_ANTENNA_INIT_PRIO, NULL);
+
+#define GPIO_DT_SPEC(n, p, i) GPIO_DT_SPEC_GET_BY_IDX(n, p, i),
```

Comment:
I think this "helper" only confuses things. Why not just use GPIO_DT_SPEC_GET_BY_IDX?

Reply:
Well, it makes more sense this way since some boards (launchxl) as all 3 gpios connected, bcf has v1 and v2 connected and beagleplay has v1 and v3.

Reply:
I mean, why use `GPIO_DT_SPEC` anywhere rather than simply use `GPIO_DT_SPEC_GET_BY_IDX`? It seems to be a needless shortcut.

Reply:
Using `GPIO_DT_SPEC` instead of `GPIO_DT_SPEC_GET_BY_IDX` means one less change for anyone wanting to copy this board antenna file and reuse it. While using IDX is not a bad thing when just one board is using this or a copy of this file. However, right now, there are at least 3 boards that use this.

Ideally, I would want to move this completely out of board if possible but right now, just taking small steps to make it as generic as possible.

Reply:
I see what you mean. The reason for introducing `GPIO_DT_SPEC` is that the macro passed to require `,` (comma). 
The reason for going with `DT_FOREACH_PROP_ELEM` instead of the original `GPIO_DT_SPEC_GET_BY_IDX_OR` (from cc1352p1_lauchxl/board_antenna.c) is that I do not think it is correct to use that macro to create an array. For example look at the following original code:
```c
static const struct gpio_dt_spec ant_gpios[] = {
	GPIO_DT_SPEC_GET_BY_IDX_OR(ANTENNA_MUX, gpios, BOARD_ANT_GPIO_24G, {0}),
	GPIO_DT_SPEC_GET_BY_IDX_OR(ANTENNA_MUX, gpios, BOARD_ANT_GPIO_PA, {0}),
	GPIO_DT_SPEC_GET_BY_IDX_OR(ANTENNA_MUX, gpios, BOARD_ANT_GPIO_SUBG, {0}),
};
```
At first glance, it might seem fine but it accesses the GPIOs in the following way below:
```c
gpio_pin_configure_dt(&ant_gpios[BOARD_ANT_GPIO_24G], 1);
```
This means `BOARD_ANT_GPIO_24G` and other constants need to be the same as the position they are placed in the array init. This means that if we move the pins around in device tree, we need to update both the array and the constant since the constant tries to define 2 completely unrelated positions.

On the other hand, when using `DT_FOREACH_PROP_ELEM`, the constant only defines the GPIO relative to device tree.

Reply:
Fwir, it's not guaranteed that the item is in the same relative order as in DT, but a few uses of that dt macro also make that assumption already.

Probably ok here.

Reply:
>  I have replaced it with DT_FOREACH_PROP_ELEM_SEP

Yes, that would be the right approach if order would not matter, see how they do it in `pinctrl_apply_state()` - there they check the ID to make order irrelevant. That's not the case for the GPIOs.

Sorry for not having seen the comma, but now that I understand what you intend, I have to say, I don't like the change...

> Fwir, it's not guaranteed that the item is in the same relative order as in DT. [...] Probably ok here.

I think there are a few reasons why the previous solution (explicit individual access to PINs) was better IMO:
* GPIOs may well be re-configured in application-specific context (which may change order). It is the intent of this change to make the file more re-usable, I think order matters here. If this breaks in an application then this ordering bug will be really hard to hunt down.
* The extra indirection/abstraction makes the code harder to follow, IMO. We are talking about 2 (BeagleConnect) and 3 (LaunchXL) pins, respectively. I think it's not worth the loss in readability (=premature abstraction).
* The number of PINs is fixed. The for-each abstraction is not required here. It is much more adequate for variable-sized arrays.

Can you revert this change in both files to make them more readable and avoid hard-to-fix bugs, please?